crates/solver/src/solver/single_order_solver.rs

```diff
@@ -274,48 +314,86 @@ impl Solver for SingleOrderSolver {
+                    match self
+                        .solve_single_order(order, &auction)
+                        .instrument(span)
+                        .await
+                    {
+                        SolveResult::Failed => continue,
+                        SolveResult::Retryable(order) => orders.push_back(order),
+                        SolveResult::Solved(settlement) => {
+                            let _ = tx.send(settlement);
+                        }
+                        SolveResult::RateLimited => {
+                            tracing::warn!(
+                                solver = %self.name(),
+                                "rate limited; backing off until next auction"
+                            );
+                            break;
+                        }
                     }
                 }
             }
-        };
-
-        // Subtract a small amount of time to ensure that the driver doesn't reach the
-        // deadline first.
-        let _ = tokio::time::timeout_at(
-            auction
-                .deadline
-                .checked_sub(Duration::from_secs(1))
-                .unwrap()
-                .into(),
-            settle,
-        )
-        .await;
+            .fuse();
+            futures::pin_mut!(solve);
+
+            let finalize = async {
+                let mut index = 0;
+                while let Some(intermediate) = rx.recv().await {
+                    match self
+                        .finalize_settlement(intermediate, &auction, index)
+                        .await
+                    {
+                        Ok(Some(settlement)) => {
+                            settlements.push(settlement);
+                        }
+                        Ok(None) => (),
+                        Err(err) => {
+                            tracing::warn!(?err, "failed to finalize intermediate settlement");
+                        }
+                    }
+                    index += 1;
+                }
+            };
+            futures::pin_mut!(finalize);
+
+            // Subtract a small amount of time to ensure that the driver doesn't reach the
+            // deadline first.
+            let timeout = tokio::time::sleep_until(
+                auction
+                    .deadline
+                    .checked_sub(Duration::from_secs(1))
+                    .unwrap()
+                    .into(),
+            );
+            futures::pin_mut!(timeout);
+
+            loop {
+                tokio::select! {
+                    // if `solve` stops early there are still settlements to be finalized
+                    _ = &mut solve => (),
+                    // all possible settlements have been finalized
+                    _ = &mut finalize => break,
+                    // we reached the timeout and should return the results ASAP
+                    _ = &mut timeout => break,
+                }
+            }
```

Comment:
I would change it like this. Then you can remove the previous borrow checker happy braces so that code is less indented.

```suggestion
        {
            let solve = solve.fuse();
            futures::pin_mut!(solve);
            futures::pin_mut!(timeout);
            futures::pin_mut!(finalize);
            loop {
                tokio::select! {
                    // if `solve` stops early there are still settlements to be finalized
                    _ = &mut solve => (),
                    // all possible settlements have been finalized
                    _ = &mut finalize => break,
                    // we reached the timeout and should return the results ASAP
                    _ = &mut timeout => break,
                }
            }
        }
```

Reply:
Neat, thanks! 👍